modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureRetryingInputStream.java

```diff
@@ -0,0 +1,89 @@
+import org.elasticsearch.common.blobstore.RetryingInputStream;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.repositories.blobstore.RequestedRangeNotSatisfiedException;
+import org.elasticsearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.nio.file.NoSuchFileException;
+
+public class AzureRetryingInputStream extends RetryingInputStream<String> {
+
+    protected AzureRetryingInputStream(AzureBlobStore azureBlobStore, OperationPurpose purpose, String blob, long position, Long length)
+        throws IOException {
+        super(
+            new AzureBlobStoreServices(azureBlobStore, purpose, blob),
+            purpose,
+            position,
+            length == null ? Long.MAX_VALUE - 1 : position + length
+        );
+    }
+
+    private record AzureBlobStoreServices(AzureBlobStore blobStore, OperationPurpose purpose, String blob)
+        implements
+            RetryingInputStream.BlobStoreServices<String> {
+
+        @Override
+        public InputStreamAtVersion<String> getInputStreamAtVersion(@Nullable String version, long start, long end) throws IOException {
+            try {
+                final Long length = end < Long.MAX_VALUE - 1 ? end - start : null;
+                final AzureBlobStore.AzureInputStream inputStream = blobStore.getInputStream(purpose, blob, start, length, version);
+                return new InputStreamAtVersion<>(inputStream, inputStream.getETag());
+            } catch (Exception e) {
+                if (ExceptionsHelper.unwrap(e, HttpResponseException.class) instanceof HttpResponseException httpResponseException) {
+                    final var httpStatusCode = httpResponseException.getResponse().getStatusCode();
+                    if (httpStatusCode == RestStatus.NOT_FOUND.getStatus()) {
+                        throw new NoSuchFileException("Blob [" + blob + "] not found");
+                    }
+                    if (httpStatusCode == RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()) {
+                        throw new RequestedRangeNotSatisfiedException(blob, start, end == Long.MAX_VALUE - 1 ? -1 : end - start, e);
+                    }
+                }
+                switch (e) {
+                    case RuntimeException runtimeException -> throw runtimeException;
+                    case IOException ioException -> throw ioException;
+                    default -> throw new IOException("Unable to get input stream for blob [" + blob + "]", e);
+                }
+            }
+        }
+
+        @Override
+        public void onRetryStarted(String action) {
+            // No metrics for Azure
+        }
+
+        @Override
+        public void onRetrySucceeded(String action, long numberOfRetries) {
+            // No metrics for Azure
+        }
+
+        @Override
+        public long getMeaningfulProgressSize() {
+            return Math.max(1L, blobStore.getReadChunkSize() / 100L);
+        }
```

Comment:
This value seems kind-of arbitrary.

The Azure value calculates to about 320KiB, the GCS to 160KiB and the S3 one to 1MiB, they are all functions of various loosely-related thresholds. Perhaps it makes sense to make this a first-class setting and consistent across the CSPs?